File: 1559-detect-cycles-in-2d-grid/1559-detect-cycles-in-2d-grid.ts
class UF {
    private root: Record<string, string>
    constructor() {
        this.root = {}
    }
    add(a: string): void {
        if (!(a in this.root)) {
            this.root[a] = a
        }
    }
    find(a: string): string {
        if (this.root[a] !== a) {
            this.root[a] = this.find(this.root[a])
        }
        return this.root[a]
    }
    union(a: string, b: string): boolean {
        const rA = this.find(a)
        const rB = this.find(b)
        if (rA === rB) return false
        this.root[rB] = rA
        return true
    }
}

function containsCycle(grid: string[][]): boolean {
    const uf = new UF()
    for (let r=0; r < grid.length; r++) {
        for (let c=0; c < grid[r].length; c++) {
            uf.add(coordToKey(r, c))
            for (const [rr, cc] of [[1, 0], [0, 1]]) {
                const rrr = rr + r
                const ccc = cc + c
                if (!(rrr in grid) || !(ccc in grid[rrr]) || grid[rrr][ccc] !== grid[r][c]) continue
                uf.add(coordToKey(rrr, ccc))
                if (uf.union(coordToKey(r, c), coordToKey(rrr, ccc)) === false) return true
            }
        }
    }
    return false
};

function coordToKey(r, c): string {
    return `${r} ${c}`
}

File: 1559-detect-cycles-in-2d-grid/1559-detect-cycles-in-2d-grid.py
class UF:
    def __init__(self):
        self.root = {}
    
    def add(self, a: str) -> None:
        if a not in self.root:
            self.root[a] = a
    
    def find(self, a: str) -> str:
        if self.root[a] != a:
            self.root[a] = self.find(self.root[a])
        return self.root[a]
    
    def union(self, a: str, b: str) -> bool:
        a, b = self.find(a), self.find(b)
        if a == b:
            return False
        self.root[b] = a

class Solution:
    def to_string(self, r: int, c: int) -> str:
        return f"{r} {c}"
    
    def containsCycle(self, grid: List[List[str]]) -> bool:
        uf = UF()
        for r in range(0, len(grid)):
            for c in range(0, len(grid[r])):
                uf.add(self.to_string(r, c))
                for rr, cc in [[r+1, c], [r, c+1]]:
                    if rr < 0 or rr >= len(grid) or cc < 0 or cc >= len(grid[rr]) or grid[rr][cc] != grid[r][c]:
                        continue
                    uf.add(self.to_string(rr, cc))
                    if uf.union(self.to_string(r, c), self.to_string(rr, cc)) == False:
                        return True
        return False